### DISCUSSION TITLE
nRF52840 Dongle Zigbee range/stability

### DISCUSSION THREAD

**Discussion by bbadrignans** [ffenix113/zigbee_home#56] [Q&A]
I use the project on several nRF52840 Dongle, they are configured as routers, have temperature/humidity sensors (so report every 3 minutes).

One dongle is quite close to my coordinator and many other routers : It runs like a charm for more than a week.
Two other dongles are far from coordinator and quite close to at least one router : They run for hours, then disappear and never come back, unless power cycled (see picture below).
I cannot get logs easily, so it is difficult to understand/debug.

![image](https://github.com/ffenix113/zigbee_home/assets/534410/c21d66a8-44b2-4663-bbde-a7e3b96dc539)

I had a zigbee power plug (From Nous manufacturer) that worked fine for week at the same places.
So I don't think the place is the cause.

Did you ever encounter the same issue with nRF52840 Dongle ?
Is there a way to enhance the dongle stability on poor signal conditions ? 

Maybe by software ? Fix a bug related to poor signal stability (In picture above, some values are missing before disappearance, I suppose the dongle should reconnect when signal is better, but it never reconnect) ? Use more power with nrf_802154_tx_power_set ? 
Maybe by hardware (better power supply, better orientation...) ?

**Comment by ffenix113** [comment]
Hmm, interesting. 

I do see similar behavior on my devices as well, so I will get this as a priority to trigger and hopefully fix.

**Comment by bbadrignans** [comment]
@ffenix113  Thank you for the release and the ias_zone fix. I will try that.
Regarding this discussion, I notice something, but I'm not really sure : 
- After some time, all my 3 dongles are now stable (run well for 1 week)
- I suppose it's because my Zigbee mesh is now stable, it had evolved after adding my 3 nRF52840 Dongle
- Recently I updated one of my router (a light bulb) and, quite at the same time, one of my nRF52840 Dongle froze 

So I suppose the crash appears when Zigbee mesh is reconfiguring/evolving.
Do you think it is possible ? 
Maybe this commit will prevent the crash : https://github.com/ffenix113/zigbee_home/pull/60
I will try the release and update this discussion if I find more clues.

**Comment by bbadrignans** [comment]
I would like to test stability without debug and #60 
However without debug my dongle does not connect to zigbee (and green led does not light up) .

Bellow my config (by the way, thank you for fixing ias_zone, it works).
FYI my 3 dongles are now stables.

```
general:
  runevery: 3m
  board: nrf52840dongle_nrf52840
  
  ncs_toolchain_base: ~/ncs/
  zephyr_base: $ZEPHYR_BASE
  ncs_version: v2.6.0

  zigbee_channels: [11,13,15,16,17]

  flasher: nrfutil
  flasheroptions:
    port: /dev/ttyACM0

  manufacturer: zh
  device_name: pool_house

board:

  debug:
    enabled: false
    console: usb
    leds:
      enabled: true
      power: led0_green
  is_router: true
  leds:
    - id: led0_green

sensors:
    - type: dht
      variant: "dht22"
      pin:
        port: 0
        pin: 31
    - type: ias_zone
      pin:
        port: 0
        pin: 2
    - type: on_off
      pin:
        port: 1
        pin: 9
        inverted: true
    - type: on_off
      pin:
        port: 0
        pin: 8
        inverted: true
```

**Reply by ffenix113** [reply]
> However without debug my dongle does not connect to zigbee

That is weird. Did you remove dongle from Zigbee network before re-flashing it and then tried to connect again, or just re-flashed while it was connected to the network?
I had re-flashed my Arduino Nano while it was connected to the network and it is working just fine. I will try with the dongle when I would have a moment for it, and probably will work on some indication of "ready-to-connect" state, together with better factory reset implementation & explanation.

**Reply by bbadrignans** [reply]
Hi @ffenix113 ,
I just re-flashed using DFU, it was previously connected to my zigbee network.
I don't even try do add it again because I though it will keep Zigbee network info. But maybe removing debug needs to add it again to the network ?
The green led was off, so I supposed the device was not working at all.
Ok I will read you explanation when available.
And yes a ready-to-connect (maybe blue led flashing) would be great.

EDIT : And for information my 3 dongles are still stable now.

**Reply by ffenix113** [reply]
> I don't even try do add it again because I though it will keep Zigbee network info. But maybe removing debug needs to add it again to the network ?

No, this should not be the case. Unless the device was wiped (or factory reset) the Zigbee network info will be stored on flash.

> Ok I will read you explanation when available.

I have added more control over factory reset button. I have not yet added doc page for it, but [example](https://github.com/ffenix113/zigbee_home/tree/a55dfc6a351b6d326f35323d3bc7c00559577909/examples/factory_reset) is present on this new feature.

Note: there were breaking changes needed to improve usability and configuration improvements, so the `contact` sensor definition has changed(though I forgot to update example, which I will do next :sweat_smile: ).

> And yes a ready-to-connect (maybe blue led flashing) would be great.

Working on this, yes. Previously there was a identify cluster, which provided necessary hardware behavior(i.e. blinking LED), but I removed it some time ago. Will add it back, and use the blinking functionality from it to show if device is in pairing mode.

> EDIT : And for information my 3 dongles are still stable now.

Good to know! Glad that it finally "worked out" :slightly_smiling_face:

**Reply by ffenix113** [reply]
Just for a quick information: some factory reset button documentation was added in this MR: https://github.com/ffenix113/zigbee_home/pull/76
